### PR TITLE
Integrating Vertex Color (For discussion only)

### PR DESCRIPTION
**Pull request by marcelluethi**
This PR shows my initial efforts to support VertexColor in Scalismo, by moving concepts from scalismo-faces into Scalismo. The PR is meant as a platform to continue the discussion started in issue #235.

In this version, the VTK PLYReader is used, instead of the one from scalismo-faces. The reason for this is that using the PLY Reader from scalismo-faces would require to port also all the concepts related to texture. If we want to do this, and whether such an integration makes sense is up for discussion.

### CODE REVIEW

File: src/main/scala/scalismo/mesh/TriangleMesh.scala [hunk 1]
```diff
@@ -199,6 +201,26 @@ object TriangleMesh3D {
   }
 }
 
+/**
+ * colored mesh with color per vertex
+ */
+case class VertexColorMesh3D(pointSet: UnstructuredPointsDomain[_3D],
```
[the remark this answers is not in the document]
Comment (marcelluethi) [reply src/main/scala/scalismo/mesh/TriangleMesh.scala]: The purpose of this exploration was to figure out how we could support VertexColor in the UI in a minimally invasive way, and without exposing any new concepts to the user. 
With our new goal of defining stable core concepts, many of the decisions will be different. I think that the design (and harmonization) of the IO methods  should be discussed along with the integration of the new mesh classes (see also issue #97). We will also need to discuss how to integrate the PLY Reader from Scalismo-faces - an issue which we also bypassed in this version.

[the remark this answers is not in the document]
Comment (Ghazi-Bouabene) [reply src/main/scala/scalismo/mesh/TriangleMesh.scala]: I also think that a composite would be more coherent with other similar constructs such as ScalarMeshField

File: src/main/scala/scalismo/io/MeshIO.scala [hunk 2]
```diff
@@ -275,6 +317,58 @@ object MeshIO {
...
 
+  private def readPLY(file: File): Try[TriangleMesh[_3D]] = {
+    val plyReader = new vtkPLYReader()
+    plyReader.SetFileName(file.getAbsolutePath)
+
+    plyReader.Update()
+
+    val errCode = plyReader.GetErrorCode()
+    if (errCode != 0) {
+      return Failure(new IOException(s"Could not read stl mesh (received error code $errCode"))
```
Comment (Ghazi-Bouabene) [inline src/main/scala/scalismo/io/MeshIO.scala]: Not that it matters for the discussion but this should say "Could not read ply mesh" (thought to write before forgetting :))

File: src/main/scala/scalismo/color/RGBA.scala [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    /** channel-wise multiplication */
+    override def multiply(pix1: RGBA, pix2: RGBA): RGBA = pix1 x pix2
+
+    /** zero element */
+    override def zero: RGBA = BlackTransparent
+
+    override val dimensionality = 4
+  }
+
+  implicit object RGBAInterpolator extends ValueInterpolator[RGBA] {
```
Comment (Ghazi-Bouabene) [inline src/main/scala/scalismo/color/RGBA.scala]: Am sure there is a viable explanation, but it is not clear to me why the interpolator (and consequently color space operations) would also have to be ported?

**Comment by Andreas-Forster** [comment]
Reopened this PR with the same content on a new branch from this repository ( see #243 ). This was mainly that people from within scalismo can easily change stuff in the branch for a more lively discussion.

**Pull request closed without merging**